driver_go18_test.go

```diff
@@ -98,6 +106,118 @@ func getQueryProcess(processes []dbProcess, query string) *dbProcess {
 	return nil
 }
 
+func pollQueryProcess(db *sql.DB, dbName, query string, totalTimeout, pollTimeout time.Duration) (*dbProcess, error) {
+	end := time.Now().Add(totalTimeout)
+	var (
+		err error
+		processList []dbProcess
+		longProcess *dbProcess
+	)
+	for time.Now().Before(end) {
+		processList, err = getProcesslist(dbName, db)
+		longProcess = getQueryProcess(processList, query)
+
+		if err != nil || longProcess == nil {
+			time.Sleep(pollTimeout)
+		} else {
+			break
+		}
+	}
+	if longProcess == nil {
+		return nil, fmt.Errorf("process for query \"%s\" not found", query)
+	}
+	return longProcess, err
+}
+
+func pollKillProcess(db *sql.DB, dbName string, processID int, totalTimeout, pollTimeout time.Duration) error {
+	end := time.Now().Add(totalTimeout)
+	var err error
+	for time.Now().Before(end) {
+		if checkProcessExists(dbName, processID, db) {
+			err = fmt.Errorf("process %d expected to be killed", processID)
+			time.Sleep(pollTimeout)
+		} else {
+			err = nil
+			break
+		}
+	}
+	return err
+}
+
+func checkCancellation(
+	db *sql.DB, dbName, query string,
+	cancel context.CancelFunc, errChan chan error,
+	getIdTimeout, killTimeout, pollTimeout time.Duration,
+) {
+	process, err := pollQueryProcess(db, dbName, query, getIdTimeout, pollTimeout)
+	if err != nil {
+		errChan <- fmt.Errorf("failed to get mysql process: %v", err)
+		return
+	}
+	cancel()
+	err = pollKillProcess(db, dbname, process.ID, killTimeout, pollTimeout)
+	if err != nil {
+		errChan <- fmt.Errorf("process %d expected to be killed", process.ID)
+		return
+	}
+	errChan <- nil
+}
+
+func waitForKill(errChan chan error, timeout time.Duration) error {
+	select {
+	case err := <-errChan:
+		if err != nil {
+			return err
+		}
+	case <-time.After(timeout):
+		fmt.Errorf("failed to kill on timeout")
+	}
+	return nil
+}
+
+func testCancel(dbt *DBTest, ctx context.Context, cancel context.CancelFunc, query string, queryFunc func() error) {
+	var err error
+	tx, err := dbt.db.BeginTx(context.Background(), nil)
+	if err != nil {
+		dbt.Fatal(err)
+	}
+
+	_, err = tx.Exec("LOCK TABLES test WRITE")
+	if err != nil {
+		tx.Rollback()
+		dbt.Fatal(err)
+	}
+
+	errChan := make(chan error)
+	go checkCancellation(dbt.db, dbname, query, cancel, errChan, getIdTimeout, killTimeout, pollTimeout)
```

Comment:
it would be better to change the logic,
execute query in separate gorutine, but kill and wait in main.

Reply:
ok